### PR TITLE
[quickfort] implement the new quickfort gui

### PR DESCRIPTION
**Pull request by myk002**
Works towards DFHack/dfhack#2018

Depends on new widgets and functions in DFHack/dfhack#2118 DFHack/dfhack#2119 DFHack/dfhack#2103

This is the first iteration of the new graphical interface for the `quickfort` script. Once you load a blueprint, you will see a blinking "shadow" over the tiles that will be modified. You can use the cursor to reposition the blueprint. Once you are satisfied, hit ENTER to apply the blueprint to the map.

Loading a blueprint automatically is implemented if you pass a filter to the `gui/quickfort` script that only matches one blueprint.

Locking the cursor is also implemented, so you can lock the blueprint in place and scan the map around to see how it would be applied offscreen.

Also implemented: previewing which orders would be generated, actually generating those orders, and blueprint undo.

After this is merged, next steps would be:
- UI for interactive blueprint transformation (e.g. rotations)
- UI for controlling repetition up or down z-levels
- asynchronous preview refreshes. the cursor can get slow when you're previewing large blueprints and moving the cursor around quickly.

### CODE REVIEW

File: internal/quickfort/query.lua [hunk 1]
```diff
@@ -16,6 +16,7 @@ local log = quickfort_common.log
 
 local function is_queryable_tile(pos)
     local flags, occupancy = dfhack.maps.getTileFlags(pos)
+    if not flags then return false end
```
Comment (myk002) [inline internal/quickfort/query.lua:19]: protect against the tile being out of map bounds (where `getTileFlags` returns `nil`)

**Comment by myk002** [comment]
Waiting to merge this until I can use the refactored widgets from `gui/blueprint`

**Review by softmoth — COMMENTED** [review]
Looks good, to my untrained eyes.

File: gui/quickfort.lua [hunk 2]
```diff
@@ -1,27 +1,498 @@
...
+BlueprintDetails.ATTRS{
+    focus_path='quickfort/dialog/details',
+    frame_title='Details',
+    frame_width=28, -- minimum width required for the bottom frame text
+}
+
+-- adds hint about left arrow being a valid "exit" key for this dialog
+function BlueprintDetails:onRenderFrame(dc, rect)
+    BlueprintDetails.super.onRenderFrame(self, dc, rect)
+    dc:seek(rect.x1+2, rect.y2):string('Left arrow', dc.cur_key_pen):
```
[the remark this answers is not in the document]
Comment (myk002) [reply gui/quickfort.lua:75]: I'll keep this in mind for later. I've been making a lot of changes to the common widget library, and this does seem like something to address later. The problem is that the standard "stringification" of the arrow key comes out as "Numpad 4" or something true, but unhelpful like that.

File: gui/quickfort.lua [hunk 3]
```diff
@@ -1,27 +1,498 @@
...
+        end
+        if v.comment then
+            full_text = full_text .. '\n\n comment: ' .. v.comment
+        end
+        local truncated_text =
+                truncate(text, self.frame_body.width, self.row_height)
+
+        -- search for the extra syntax shown in the list items in case someone
+        -- is typing exactly what they see
+        local extra_keys = main
```
[the remark this answers is not in the document]
Comment (myk002) [reply gui/quickfort.lua:207]: indeed. an artifact from a previous implementation. thanks!

File: gui/quickfort.lua [hunk 4]
```diff
@@ -1,27 +1,498 @@
...
+will see a blinking "shadow" over the tiles that will be modified. You can use
+the cursor to reposition the blueprint, or use the hotkeys to rotate or flip the
+blueprint around to your liking. Once you are satisfied, hit :kbd:`ENTER` to
+apply the blueprint to the map.
+
+Usage::
+
+    gui/quickfort [<search terms>]
+
+If the specified search terms match exactly one blueprint (e.g. if the search
```
[the remark this answers is not in the document]
Comment (myk002) [reply gui/quickfort.lua:16]: Updated. PTAL.

**Pull request merged**